src/libsystemd/sd-device/sd-device.c

```diff
@@ -1180,21 +1180,14 @@ static int device_set_sysname_and_sysnum(sd_device *device) {
                 return r;
 
         /* some devices have '!' in their name, change that to '/' */
-        while (sysname[len] != '\0') {
-                if (sysname[len] == '!')
-                        sysname[len] = '/';
-
-                len++;
-        }
-
-        /* trailing number */
-        while (len > 0 && isdigit(sysname[--len]))
-                sysnum = &sysname[len];
+        for (p = strchrnul(sysname, '!'); *p != '\0'; p = strchrnul(p, '!'))
+                *p = '/';
 
-        if (len == 0)
-                sysnum = NULL;
+        /* trailing number (refuse number only sysname)*/
+        for (; p > sysname && isdigit(p[-1]); p--)
```

Comment:
btw, isdigit() is kinda nasty, since in theory locale dependent (i.e. there's even `isdigt_l()` to run it with a different locale!). To be locale-independent we should probably replace most cases where we use `isdigit(x)` right now by `strchr(DIGITS, x)` instead. Since most of the time we only care about ASCII, like in this case.